File: android/app/src/main/java/com/flutcloud/flutlink/data/FlutCloudApi.kt
package com.flutcloud.flutlink.data

import com.flutcloud.flutlink.data.dto.AppInfoDto
import com.flutcloud.flutlink.data.dto.FlutCloudItemDto
import com.flutcloud.flutlink.data.dto.ManagedUser
import com.flutcloud.flutlink.data.dto.Quota
import com.flutcloud.flutlink.data.dto.QuotaDto
import com.flutcloud.flutlink.data.dto.SessionUser
import com.flutcloud.flutlink.data.dto.Share
import com.flutcloud.flutlink.data.dto.ShareDto
import com.flutcloud.flutlink.data.dto.UserDetailsDto
import com.flutcloud.flutlink.data.dto.UserInfoDto
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.Credentials
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException
import java.net.URLEncoder

/**
 * OCS + FlutCloud-app API client. All HTTP traffic runs on OkHttp; every
 * request is verified against the `flutcloud` capability before an account
 * is accepted (FlutCloud-only policy).
 */
class FlutCloudApi(private val client: OkHttpClient) {

    private val json = Json { ignoreUnknownKeys = true }
    private val tag = "FlutLinkOcs"

    /** Parse the OCS meta + return `(data, errorMessage)`. */
    private fun parseOcs(body: String): Pair<JsonElement?, String?> {
        val obj = runCatching { json.parseToJsonElement(body).jsonObject }.getOrNull()
            ?: return null to "Invalid OCS response"
        val meta = obj["ocs"]?.jsonObject?.get("meta")?.jsonObject
        val status = (meta?.get("status") as? JsonPrimitive)?.contentOrNull
        val code = (meta?.get("statuscode") as? JsonPrimitive)?.contentOrNull
        val message = (meta?.get("message") as? JsonPrimitive)?.contentOrNull
            ?: "Unknown OCS error"
        val ok = status.equals("ok", ignoreCase = true) || code == "100" || code == "200"
        return obj["ocs"]?.jsonObject?.get("data") to if (ok) null else message
    }

    private fun build(
        session: AuthSession,
        method: String,
        url: String,
        form: List<Pair<String, String>>? = null
    ): Request {
        val builder = Request.Builder()
            .url(url)
            .method(method, form?.let { pairs ->
                FormBody.Builder().apply { pairs.forEach { (k, v) -> add(k, v) } }.build()
            })
            .header("Authorization", Credentials.basic(session.username, session.token))
            .header("OCS-APIRequest", "true")
            .header("Accept", "application/json")
        return builder.build()
    }

    /** Execute an OCS request; throws [ApiException] on HTTP/meta errors. */
    private suspend fun execute(
        session: AuthSession,
        method: String,
        url: String,
        form: List<Pair<String, String>>? = null
    ): JsonElement? = withContext(Dispatchers.IO) {
        val request = build(session, method, url, form)
        val started = System.currentTimeMillis()
        try {
            client.newCall(request).execute().use { response ->
                val body = response.body?.string().orEmpty()
                android.util.Log.i(tag, "$method $url -> ${response.code} in ${System.currentTimeMillis() - started}ms body=${body.length}")
                if (response.code >= 400) {
                    throw ApiException(
                        "Server answered ${response.code}: $body".trim(),
                        "http_${response.code}",
                        response.code
                    )
                }
                val (data, error) = parseOcs(body)
                if (error != null) throw ApiException(error, "ocs_error", response.code)
                data
            }
        } catch (e: ApiException) {
            throw e
        } catch (e: IOException) {
            throw NetworkException(e)
        }
    }

    private fun String.encoded() = URLEncoder.encode(this, "UTF-8").replace("+", "%20")

    // --- FlutCloud-only verification --------------------------------------

    /**
     * Verify that the server runs the FlutCloud app. Queries the OCS
     * capabilities endpoint **with the account credentials** — Nextcloud only
     * includes app capabilities (`ocs.data.capabilities.flutcloud`) for
     * authenticated requests, so an anonymous probe would reject valid
     * servers. Mirrors the desktop client's `verify_server`.
     */
    suspend fun verifyServer(session: AuthSession) {
        val url = "${session.normalizedBaseUrl}/ocs/v2.php/cloud/capabilities?format=json"
        val data = execute(session, "GET", url) ?: throw FlutCloudAppMissing()
        val has = data.jsonObject["capabilities"]
            ?.jsonObject
            ?.containsKey("flutcloud") == true
        if (!has) throw FlutCloudAppMissing()
    }

    // --- Current user / session -------------------------------------------

    suspend fun getCurrentUser(session: AuthSession): SessionUser {
        val data = execute(session, "GET", "${session.normalizedBaseUrl}/ocs/v2.php/cloud/user?format=json")
            ?: throw ApiException("Missing user data")
        val user = json.decodeFromString<UserInfoDto>(data.toString())
        return SessionUser(
            id = user.id ?: throw ApiException("Missing user id"),
            displayName = user.displayName,
            isAdmin = user.isAdmin ?: false
        )
    }

    suspend fun getCurrentQuota(session: AuthSession): Quota? {
        val data = execute(session, "GET", "${session.normalizedBaseUrl}/ocs/v2.php/cloud/user?format=json")
            ?: return null
        return json.decodeFromString<UserInfoDto>(data.toString()).quota?.toQuota()
    }

    /**
     * Probe admin rights. OCS v1 answers HTTP 200 even for denied requests,
     * so the result is judged from the OCS meta (`statuscode`), not the HTTP
     * status. Returns `false` when the server answered and denied the request;
     * network/parse failures propagate so callers can keep the previously
     * stored flag instead of demoting an admin account on a transient error.
     */
    suspend fun isAdmin(session: AuthSession): Boolean = withContext(Dispatchers.IO) {
        val request = build(session, "GET", "${session.normalizedBaseUrl}/ocs/v1.php/cloud/users?format=json&limit=1")
        try {
            client.newCall(request).execute().use { response ->
                val body = response.body?.string().orEmpty()
                if (response.code >= 400) {
                    throw ApiException(
                        "Server answered ${response.code}: $body".trim(),
                        "http_${response.code}",
                        response.code
                    )
                }
                val (_, error) = parseOcs(body)
                error == null
            }
        } catch (e: ApiException) {
            throw e
        } catch (e: IOException) {
            throw NetworkException(e)
        }
    }

    suspend fun ping(session: AuthSession): AppInfoDto? {
        val data = execute(session, "GET", "${session.normalizedBaseUrl}/ocs/v2.php/apps/flutcloud/api/v1/ping")
            ?: return null
        return json.decodeFromString<AppInfoDto>(data.toString())
    }

    // --- OCS Provisioning API ---------------------------------------------

    suspend fun listUsers(session: AuthSession, search: String = ""): List<String> {
        val limit = 200
        val all = mutableListOf<String>()
        val seen = mutableSetOf<String>()
        var offset = 0
        while (true) {
            var url = "${session.normalizedBaseUrl}/ocs/v1.php/cloud/users?format=json&limit=$limit&offset=$offset"
            if (search.isNotEmpty()) url += "&search=${search.encoded()}"
            val data = execute(session, "GET", url) ?: break
            val users = data.jsonObject["users"]?.jsonArray
                ?.mapNotNull { it.jsonPrimitive.contentOrNull }
                ?.filter { seen.add(it) }
                .orEmpty()
            if (users.isEmpty()) break
            all += users
            if (users.size < limit) break
            offset += limit
        }
        return all
    }

    suspend fun getUser(session: AuthSession, userId: String): ManagedUser {
        val data = execute(
            session, "GET",
            "${session.normalizedBaseUrl}/ocs/v1.php/cloud/users/${userId.encoded()}?format=json"
        ) ?: throw ApiException("Missing user data")
        val dto = json.decodeFromString<UserDetailsDto>(data.toString())
        return ManagedUser(
            id = dto.id ?: userId,
            displayName = dto.displayName,
            email = dto.email,
            quota = dto.quota?.toQuota(),
            groups = dto.groups.orEmpty(),
            enabled = dto.enabled ?: true
        )
    }

    suspend fun createUser(session: AuthSession, userId: String, password: String, displayName: String? = null) {
        val form = mutableListOf("userid" to userId, "password" to password)
        if (!displayName.isNullOrEmpty()) form += "displayName" to displayName
        execute(session, "POST", "${session.normalizedBaseUrl}/ocs/v1.php/cloud/users?format=json", form)
    }

    suspend fun deleteUser(session: AuthSession, userId: String) {
        execute(session, "DELETE", "${session.normalizedBaseUrl}/ocs/v1.php/cloud/users/${userId.encoded()}?format=json")
    }

    suspend fun updateUser(session: AuthSession, userId: String, key: String, value: String) {
        execute(
            session, "PUT",
            "${session.normalizedBaseUrl}/ocs/v1.php/cloud/users/${userId.encoded()}?format=json",
            listOf("key" to key, "value" to value)
        )
    }

    suspend fun setUserQuota(session: AuthSession, userId: String, quotaBytes: Long?) {
        val value = quotaBytes?.toString() ?: "none"
        updateUser(session, userId, "quota", value)
    }

    // --- OCS files sharing API --------------------------------------------

    suspend fun createShare(
        session: AuthSession,
        path: String,
        shareType: Int,
        shareWith: String? = null,
        password: String? = null,
        expireDate: String? = null,
        publicUpload: Boolean = false,
        permissions: Long? = null
    ): Share {
        val form = mutableListOf("path" to path, "shareType" to shareType.toString())
        shareWith?.let { form += "shareWith" to it }
        password?.let { form += "password" to it }
        expireDate?.let { form += "expireDate" to it }
        val perms = permissions ?: when {
            publicUpload -> 15
            shareType == 3 -> 1
            else -> null
        }
        perms?.let { form += "permissions" to it.toString() }
        val data = execute(
            session, "POST",
            "${session.normalizedBaseUrl}/ocs/v2.php/apps/files_sharing/api/v1/shares?format=json",
            form
        ) ?: throw ApiException("Share endpoint returned no data")
        return json.decodeFromString<ShareDto>(data.toString()).toShare()
            ?: throw ApiException("Share endpoint returned no share data")
    }

    suspend fun listShares(session: AuthSession, path: String? = null): List<Share> {
        var url = "${session.normalizedBaseUrl}/ocs/v2.php/apps/files_sharing/api/v1/shares?format=json"
        path?.let { url += "&path=${it.encoded()}" }
        val data = execute(session, "GET", url) ?: return emptyList()
        return data.jsonArray.mapNotNull { json.decodeFromString<ShareDto>(it.toString()).toShare() }
    }

    suspend fun deleteShare(session: AuthSession, shareId: Long) {
        execute(session, "DELETE", "${session.normalizedBaseUrl}/ocs/v2.php/apps/files_sharing/api/v1/shares/$shareId?format=json")
    }

    // --- FlutCloud app: virtual links + writable parts --------------------

    suspend fun listLinks(session: AuthSession): List<FlutCloudItemDto> {
        val data = execute(session, "GET", "${session.normalizedBaseUrl}/ocs/v2.php/apps/flutcloud/api/v1/links")
            ?: return emptyList()
        return data.jsonArray.map { json.decodeFromString<FlutCloudItemDto>(it.toString()) }
    }

    suspend fun createLink(session: AuthSession, name: String) {
        execute(
            session, "POST",
            "${session.normalizedBaseUrl}/ocs/v2.php/apps/flutcloud/api/v1/links",
            listOf("name" to name)
        )
    }

    suspend fun deleteLink(session: AuthSession, name: String) {
        execute(session, "DELETE", "${session.normalizedBaseUrl}/ocs/v2.php/apps/flutcloud/api/v1/links/${name.encoded()}")
    }

    suspend fun listParts(session: AuthSession): List<FlutCloudItemDto> {
        val data = execute(session, "GET", "${session.normalizedBaseUrl}/ocs/v2.php/apps/flutcloud/api/v1/parts")
            ?: return emptyList()
        return data.jsonArray.map { json.decodeFromString<FlutCloudItemDto>(it.toString()) }
    }

    suspend fun createPart(session: AuthSession, name: String) {
        execute(
            session, "POST",
            "${session.normalizedBaseUrl}/ocs/v2.php/apps/flutcloud/api/v1/parts",
            listOf("name" to name)
        )
    }
}

private fun QuotaDto.toQuota(): Quota = Quota(
    total = totalBytes,
    used = usedBytes,
    free = freeBytes,
    relative = relativePercent
)

private fun ShareDto.toShare(): Share? {
    val id = id ?: return null
    return Share(
        id = id,
        shareType = shareType?.toInt() ?: 0,
        path = path,
        shareWith = shareWith,
        shareWithDisplayName = shareWithDisplayName,
        permissions = permissions,
        url = url,
        hasPassword = hasPassword,
        expiration = expiration
    )
}

File: ios/FlutLink/Data/Networking/FlutCloudApi.swift
// TEST — iOS port created by opencode. Not a production build.
import Foundation

final class FlutCloudApi {
    private let session: AuthSession

    init(session: AuthSession) {
        self.session = session
    }

    func withSession(_ session: AuthSession) -> FlutCloudApi {
        FlutCloudApi(session: session)
    }

    // MARK: - OCS Helpers

    private func parseOcs(_ data: Data) throws -> (AnyCodable?, String?) {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let ocs = json["ocs"] as? [String: Any],
              let meta = ocs["meta"] as? [String: Any] else {
            return (nil, "Invalid OCS response")
        }
        let status = meta["status"] as? String
        let code = meta["statuscode"] as? String ?? "\(meta["statuscode"] ?? "")"
        let message = meta["message"] as? String ?? "Unknown OCS error"
        let ok = status?.lowercased() == "ok" || code == "100" || code == "200"
        let ocsData = ocs["data"]
        let wrappedData: AnyCodable? = ocsData != nil ? AnyCodable(ocsData) : nil
        return (wrappedData, ok ? nil : message)
    }

    private func executeOcs(
        method: String,
        path: String,
        formFields: [(String, String)]? = nil
    ) async throws -> AnyCodable {
        let url = URL(string: "\(session.normalizedBaseUrl)\(path)")!
        var request = HttpClient.ocsApiRequest(url: url, method: method, session: session, formFields: formFields)
        let (data, response) = try await HttpClient.perform(request)
        let (ocsData, ocsError) = try parseOcs(data)
        if response.statusCode >= 400 {
            let body = String(data: data, encoding: .utf8) ?? ""
            throw ApiException.api(message: ocsError ?? "Server answered \(response.statusCode): \(body)", code: "http_\(response.statusCode)", statusCode: response.statusCode)
        }
        if let err = ocsError {
            throw ApiException.api(message: err, code: "ocs_error", statusCode: response.statusCode)
        }
        guard let result = ocsData else {
            throw ApiException.api(message: "Missing OCS data")
        }
        return result
    }

    // MARK: - FlutCloud Verification

    func verifyServer() async throws {
        let data = try await executeOcs(method: "GET", path: "/ocs/v2.php/cloud/capabilities?format=json")
        let has = data.value as? [String: Any]
        let caps = has?["capabilities"] as? [String: Any]
        guard caps?["flutcloud"] != nil else {
            throw ApiException.flutCloudAppMissing
        }
    }

    // MARK: - Current User

    func getCurrentUser() async throws -> SessionUser {
        let data = try await executeOcs(method: "GET", path: "/ocs/v2.php/cloud/user?format=json")
        guard let dict = data.value as? [String: Any] else {
            throw ApiException.api(message: "Missing user data")
        }
        guard let id = dict["id"] as? String else {
            throw ApiException.api(message: "Missing user id")
        }
        return SessionUser(
            id: id,
            displayName: dict["display-name"] as? String,
            isAdmin: dict["isAdmin"] as? Bool ?? false
        )
    }

    func getCurrentQuota() async -> Quota? {
        guard let data = try? await executeOcs(method: "GET", path: "/ocs/v2.php/cloud/user?format=json"),
              let dict = data.value as? [String: Any],
              let quotaDict = dict["quota"] as? [String: Any] else {
            return nil
        }
        let qd = AnyCodable(quotaDict)
        return Quota(
            total: qd.totalBytes,
            used: qd.usedBytes,
            free: qd.freeBytes,
            relative: qd.relativePercent
        )
    }

    func isAdmin() async -> Bool {
        do {
            _ = try await executeOcs(method: "GET", path: "/ocs/v1.php/cloud/users?format=json&limit=1")
            return true
        } catch let error as ApiException {
            if error.statusCode == 401 || error.statusCode == 403 || error.code == "ocs_error" {
                return false
            }
            return false
        } catch {
            return false
        }
    }

    func ping() async -> AppInfoDto? {
        guard let data = try? await executeOcs(method: "GET", path: "/ocs/v2.php/apps/flutcloud/api/v1/ping") else {
            return nil
        }
        return try? JSONDecoder().decode(AppInfoDto.self, from: JSONSerialization.data(withJSONObject: data.value ?? [:]))
    }

    // MARK: - OCS Provisioning API

    func listUsersPage(search: String, offset: Int, limit: Int = 200) async throws -> [String] {
        var path = "/ocs/v1.php/cloud/users?format=json&limit=\(limit)&offset=\(offset)"
        if !search.isEmpty { path += "&search=\(search.urlEncoded)" }
        let data = try await executeOcs(method: "GET", path: path)
        guard let dict = data.value as? [String: Any],
              let users = dict["users"] as? [String] else {
            return []
        }
        return users
    }

    func getUser(userId: String) async throws -> ManagedUser {
        let data = try await executeOcs(method: "GET", path: "/ocs/v1.php/cloud/users/\(userId.urlEncoded)?format=json")
        guard let dict = data.value as? [String: Any] else {
            throw ApiException.api(message: "Missing user data")
        }
        let quotaDict = dict["quota"] as? [String: Any]
        let qd = quotaDict.map { AnyCodable($0) }
        return ManagedUser(
            id: dict["id"] as? String ?? userId,
            displayName: dict["display-name"] as? String,
            email: dict["email"] as? String,
            quota: qd.flatMap { Quota(total: $0.totalBytes, used: $0.usedBytes, free: $0.freeBytes, relative: $0.relativePercent) },
            groups: dict["groups"] as? [String] ?? [],
            enabled: dict["enabled"] as? Bool ?? true
        )
    }

    func createUser(userId: String, password: String, displayName: String? = nil) async throws {
        var fields: [(String, String)] = [("userid", userId), ("password", password)]
        if let dn = displayName, !dn.isEmpty { fields.append(("displayName", dn)) }
        _ = try await executeOcs(method: "POST", path: "/ocs/v1.php/cloud/users?format=json", formFields: fields)
    }

    func deleteUser(userId: String) async throws {
        _ = try await executeOcs(method: "DELETE", path: "/ocs/v1.php/cloud/users/\(userId.urlEncoded)?format=json")
    }

    func updateUser(userId: String, key: String, value: String) async throws {
        _ = try await executeOcs(method: "PUT", path: "/ocs/v1.php/cloud/users/\(userId.urlEncoded)?format=json", formFields: [("key", key), ("value", value)])
    }

    func setUserQuota(userId: String, quotaBytes: Int64?) async throws {
        let value = quotaBytes?.description ?? "none"
        try await updateUser(userId: userId, key: "quota", value: value)
    }

    // MARK: - Groups

    func listGroups(search: String = "") async throws -> [String] {
        let limit = 200
        var all: [String] = []
        var seen = Set<String>()
        var offset = 0
        while true {
            var path = "/ocs/v1.php/cloud/groups?format=json&limit=\(limit)&offset=\(offset)"
            if !search.isEmpty { path += "&search=\(search.urlEncoded)" }
            let data = try await executeOcs(method: "GET", path: path)
            guard let dict = data.value as? [String: Any],
                  let groups = dict["groups"] as? [String] else { break }
            let newGroups = groups.filter { seen.insert($0).inserted }
            if newGroups.isEmpty { break }
            all.append(contentsOf: newGroups)
            if newGroups.count < limit { break }
            offset += limit
        }
        return all
    }

    func createGroup(groupId: String) async throws {
        _ = try await executeOcs(method: "POST", path: "/ocs/v1.php/cloud/groups?format=json", formFields: [("groupid", groupId)])
    }

    func addGroupMember(groupId: String, userId: String) async throws {
        _ = try await executeOcs(method: "POST", path: "/ocs/v1.php/cloud/groups/\(groupId.urlEncoded)?format=json", formFields: [("userid", userId)])
    }

    func removeGroupMember(groupId: String, userId: String) async throws {
        _ = try await executeOcs(method: "DELETE", path: "/ocs/v1.php/cloud/groups/\(groupId.urlEncoded)/users/\(userId.urlEncoded)?format=json")
    }

    // MARK: - Shares

    func createShare(
        path: String,
        shareType: Int,
        shareWith: String? = nil,
        password: String? = nil,
        expireDate: String? = nil,
        publicUpload: Bool = false,
        permissions: Int64? = nil
    ) async throws -> Share {
        var fields: [(String, String)] = [("path", path), ("shareType", "\(shareType)")]
        if let sw = shareWith { fields.append(("shareWith", sw)) }
        if let pw = password, !pw.isEmpty { fields.append(("password", pw)) }
        if let ed = expireDate, !ed.isEmpty { fields.append(("expireDate", ed)) }
        let perms: Int64? = permissions ?? (publicUpload ? 15 : (shareType == 3 ? 1 : nil))
        if let p = perms { fields.append(("permissions", "\(p)")) }
        let data = try await executeOcs(method: "POST", path: "/ocs/v2.php/apps/files_sharing/api/v1/shares?format=json", formFields: fields)
        guard let dict = data.value as? [String: Any],
              let share = try? JSONDecoder().decode(ShareDto.self, from: JSONSerialization.data(withJSONObject: dict)),
              let id = share.id else {
            throw ApiException.api(message: "Share endpoint returned no share data")
        }
        return Share(
            id: id,
            shareType: Int(share.shareType ?? 0),
            path: share.path,
            shareWith: share.shareWith,
            shareWithDisplayName: share.shareWithDisplayName,
            permissions: share.permissions,
            url: share.url,
            hasPassword: share.hasPassword,
            expiration: share.expiration
        )
    }

    func listShares(path: String? = nil) async throws -> [Share] {
        var urlPath = "/ocs/v2.php/apps/files_sharing/api/v1/shares?format=json"
        if let p = path { urlPath += "&path=\(p.urlEncoded)" }
        let data = try await executeOcs(method: "GET", path: urlPath)
        guard let arr = data.value as? [[String: Any]] else { return [] }
        return arr.compactMap { dict in
            guard let dto = try? JSONDecoder().decode(ShareDto.self, from: JSONSerialization.data(withJSONObject: dict)),
                  let id = dto.id else { return nil }
            return Share(
                id: id,
                shareType: Int(dto.shareType ?? 0),
                path: dto.path,
                shareWith: dto.shareWith,
                shareWithDisplayName: dto.shareWithDisplayName,
                permissions: dto.permissions,
                url: dto.url,
                hasPassword: dto.hasPassword,
                expiration: dto.expiration
            )
        }
    }

    func deleteShare(shareId: Int64) async throws {
        _ = try await executeOcs(method: "DELETE", path: "/ocs/v2.php/apps/files_sharing/api/v1/shares/\(shareId)?format=json")
    }

    // MARK: - Virtual Links & Parts

    func listLinks() async throws -> [FlutCloudItemDto] {
        let data = try await executeOcs(method: "GET", path: "/ocs/v2.php/apps/flutcloud/api/v1/links")
        guard let arr = data.value as? [[String: Any]] else { return [] }
        return arr.compactMap { try? JSONDecoder().decode(FlutCloudItemDto.self, from: JSONSerialization.data(withJSONObject: $0)) }
    }

    func createLink(name: String) async throws {
        _ = try await executeOcs(method: "POST", path: "/ocs/v2.php/apps/flutcloud/api/v1/links", formFields: [("name", name)])
    }

    func deleteLink(name: String) async throws {
        _ = try await executeOcs(method: "DELETE", path: "/ocs/v2.php/apps/flutcloud/api/v1/links/\(name.urlEncoded)")
    }

    func listParts() async throws -> [FlutCloudItemDto] {
        let data = try await executeOcs(method: "GET", path: "/ocs/v2.php/apps/flutcloud/api/v1/parts")
        guard let arr = data.value as? [[String: Any]] else { return [] }
        return arr.compactMap { try? JSONDecoder().decode(FlutCloudItemDto.self, from: JSONSerialization.data(withJSONObject: $0)) }
    }

    func createPart(name: String) async throws {
        _ = try await executeOcs(method: "POST", path: "/ocs/v2.php/apps/flutcloud/api/v1/parts", formFields: [("name", name)])
    }
}
